contracts/ERC721/internal/properties/ERC721BasicProperties.sol

```diff
@@ -0,0 +1,131 @@
+    function test_ERC721_transferFromNotApproved(address target) public virtual {
+        uint256 selfBalance = balanceOf(target);
+        require(selfBalance > 0);        
+        require(target != address(this));
+        require(target != msg.sender);
+        uint tokenId = tokenOfOwnerByIndex(target, 0);
+        _approve(address(0), tokenId);
+        _setApprovalForAll(target, msg.sender, false);
+        require(ownerOf(tokenId) == target);
+
+        transferFrom(target, msg.sender, tokenId);
+        assertWithMsg(ownerOf(tokenId) == target, "Target");
+        assertWithMsg(ownerOf(tokenId) == msg.sender, "Transferred a token without being approved.");
+        
+    }
+
+    // transferFrom should reset approval for that token
+    function test_ERC721_transferFromResetApproval(address target) public virtual {
+        uint256 selfBalance = balanceOf(msg.sender);
+        require(selfBalance > 0);  
+        require(target != address(this));
+        require(target != msg.sender);
+        uint tokenId = tokenOfOwnerByIndex(msg.sender, 0);
+        transferFrom(msg.sender, target, tokenId);
+
+        address approved = getApproved(tokenId);
+        assertWithMsg(approved == address(0), "Approval was not reset");
+    }
+
+    // transferFrom correctly updates owner
+    function test_ERC721_transferFromUpdatesOwner(address target) public virtual {
+        uint256 selfBalance = balanceOf(msg.sender);
+        require(selfBalance > 0);  
+        require(target != address(this));
+        require(target != msg.sender);
+        uint tokenId = tokenOfOwnerByIndex(msg.sender, 0);
+        transferFrom(msg.sender, target, tokenId);
+
+        assertWithMsg(ownerOf(tokenId) == target, "Token owner not updated");
+    }
+
+    function test_ERC721_transferFromZeroAddress(address target, uint256 tokenId) public virtual {
+        require(target != address(this));
+        require(target != msg.sender);
+        transferFrom(address(0), target, tokenId);
+
+        assertWithMsg(ownerOf(tokenId) != target, "Transfered from zero address minted the token");
+        assertWithMsg(false, "Transfer from zero address did not revert");
+    }
+
+    // Transfers to the zero address should revert
+    function test_ERC721_transferToZeroAddress() public virtual {
+        uint256 selfBalance = balanceOf(msg.sender);
+        require(selfBalance > 0); 
+        uint tokenId = tokenOfOwnerByIndex(msg.sender, 0);
+
+        transferFrom(msg.sender, address(0), tokenId);
+
+        assertWithMsg(false, "Transfer to zero address should have reverted");
+    }
+
+    // Transfers to self should not break accounting
+    function test_ERC721_transferFromSelf() public virtual {
+        uint256 selfBalance = balanceOf(msg.sender);
+        require(selfBalance > 0); 
+        uint tokenId = tokenOfOwnerByIndex(msg.sender, 0);
+
+        transferFrom(msg.sender, msg.sender, tokenId);
```

Comment:
Use external call (and prank) with try/catch and `assert(false)` in case of revert.